### ISSUE TITLE
Lasagne/Theano model running much slower than the same model using Caffe

### ISSUE THREAD

**Issue by kli-nlpr** [Lasagne/Lasagne#663]
I used to use Caffe for my research. Now I want to use Theano based framework since it is very flexible. I choose to use Lasagne now, it is a wonderful library. But one problem bothers me very long time, the running speed of Lasagne is much slower than Caffe. 

I will show this problem by one of my own project which use AlexNet to do regression. The output layer is a scalar and I use Mean Absolute Error as loss function.
I define the network using Lasagne which is similar to the one in Lasagne’s model zoo (https://github.com/Lasagne/Recipes/blob/master/modelzoo/caffe_reference.py).

I use the same learning rate, weight decay, momentum as the Caffe model.

I define my ```iterate_minibatches``` function to read a batch of images and labels from a text file (one line of this text file contains the path of image and label) which is similar to the function at https://github.com/Lasagne/Recipes/blob/master/papers/deep_residual_learning/Deep_Residual_Learning_CIFAR-10.py#L157

Note that, for Caffe I also define a python data layer (without preloading) which is nearly the same as ```iterate_minibatches```, so I think the data loading and preprocessing time is the same for Caffe and Lasagne code. 

But the training speed is very different, 20 iterations take 12 seconds for Caffe, while it takes 46 seconds for Lasagne.  

The code of my Lasagne model definition and tranining procedure at here  https://gist.github.com/kli-nlpr/41ec0eb567b09f29299958e9d3455eef

The training log of Lasagne model at here https://gist.github.com/kli-nlpr/14d94398d9157fafb0ba5ec778c7d536

The training log of Caffe model at here https://gist.github.com/kli-nlpr/3d57c9797538ed15e3647327f644a043

The profile of Lasagne code at here https://gist.github.com/kli-nlpr/4fd1d6cf2a90ff584fe4177c43e2991a

I really appreciate your help and suggestions, Thanks.

**Comment by benanne** [comment]
I didn't have time to look into this in detail, but I wouldn't be surprised if the `LocalResponseNormalization2DLayer`s are the culprit. The implementation of these provided with Lasagne is very inefficient. We never bothered to optimise this because basically nobody uses them anymore (except for reproducing older nets like AlexNet). If you're retraining from scratch, I would suggest omitting them. Alex Krizhevsky only ever reported very minor improvements using them, and later people (informally) reported that they don't make a difference if you use a better initialization strategy (taking into account the fan-in / fan-out of the units, which is something Lasagne does by default).

**Comment by kli-nlpr** [comment]
Hi, @benanne 

Thank you very much, I tried as you said and remove the ```LocalResponseNormalization2DLayer```s.
The running speed become twice as fast. Now the time for 20 iterations become 27 seconds compared to the 46 seconds with ```LocalResponseNormalization2DLayer```.

You said that ‘it is a good idea to amortise this step by loading a large chunk of data into GPU at once’

Is there some code to achieve this?  I think this is a general problem when someone use Theano based framework for training on large scale image data set.

For example, I have a large text file which contains the image path and labels, how to effectively do batch training on this kind of dataset using Lasagne? I searched on Google but found nothing helpful…

I really appreciate your help and suggestions, Thanks.

**Comment by benanne** [comment]
I can't think of an educative example to point you to off the top of my head, sorry. This code uses this strategy: https://github.com/benanne/kaggle-ndsb/blob/master/train_convnet.py
... but it's not exactly readable.
If I recall correctly, the Theano tutorial (MNIST etc.) loads the entire dataset into a shared variable. Maybe you can use that code as a basis. Then it just comes down to periodically refreshing the contents of said shared variable.

**Comment by kli-nlpr** [comment]
Thank you very much, @benanne   I will check it out~

**Comment by kli-nlpr** [comment]
@benanne , I found your kaggle project super useful.

If only I could find this project earlier.
